substrate/frame/support/src/storage/state_diff_guard.rs

```diff
@@ -0,0 +1,381 @@
+//! 	#[pallet::storage]
+//! 	pub type SomeMap<T> = StorageMap<_, Twox64Concat, u32, u32>;
+//!     
+//!     #[pallet::storage]
+//! 	pub type SomeDoubleMap<T> = StorageMap<_, Twox64Concat, u32, u32>;
+//! }
+//!
+//! mod migrations {
+//! 	use super::*;
+//!
+//!     pub struct UncheckedMigrateToV1<T: crate::Config>(sp_std::marker::PhantomData<T>);
+//!
+//!     impl<T: crate::Config> frame_support::traits::UncheckedOnRuntimeUpgrade for UncheckedMigrateToV1<T> {
+//!         fn on_runtime_upgrade() -> frame_support::weights::Weight {
+//! 		   // migration logic here
+//!            Weight::default()
+//!         }
+//!         
+//!         #[cfg(feature = "try-runtime")]
+//!         fn try_on_runtime_upgrade() -> Result<frame_support::weights::Weight, &'static str> {
+//! 		   // migration logic here
+//!           let guard = StateDiffGuard::builder()
+//! 			.must_change(StoragePrefix {
+//! 				pallet_name: SomeMap::<T>::pallet_prefix(),
+//! 				storage_name: SomeMap::<T>::storage_prefix(),
+//! 			})
+//! 			.must_change(StoragePrefix {
+//! 				pallet_name: SomeDoubleMap::<T>::pallet_prefix(),
+//! 				storage_name: SomeDoubleMap::<T>::storage_prefix(),
+//! 			})
+//! 			.build();
+//!
+//!           // try runtime upgrade logic here
+//! 		  let weight = Self::on_runtime_upgrade();
+//!           
+//!           // any other logic
+//! 		  Ok(weight)
+//! 	   }
+//!     }
+//!
+//!     pub type MigrateToV1<T> = frame_support::migrations::VersionedMigration<
+//!     	1,
+//!      	2,
+//!      	UncheckedMigrateToV1<T>,
+//!      	Pallet<T>,
+//! 		<T as frame_system::Config>::DbWeight
+//!     >;
+//! }
+//! ```
+//!
+//! In the example above, the guard will panic if any storage entry that doesn't match `SomeMap` or `SomeDoubleMap`
+//! prefixes is changed.
+
+use core::fmt::{Debug, Formatter};
+use sp_std::{
+	collections::{btree_map::BTreeMap, btree_set::BTreeSet},
+	vec::Vec,
+};
+
+use super::storage_prefix;
+
+/// Raw storage key.
+type StorageKey = Vec<u8>;
+/// Raw storage value.
+type StorageValue = Vec<u8>;
+
+/// Storage prefix: pallet name and storage name.
+#[derive(Clone, Copy, Ord, PartialOrd, Eq, PartialEq)]
+pub struct StoragePrefix {
+	/// Name of the pallet.
+	pub pallet_name: &'static [u8],
+	/// Name of the storage.
+	pub storage_name: &'static [u8],
+}
+
+impl StoragePrefix {
+	/// Computes the storage prefix.
+	pub fn storage_prefix(&self) -> [u8; 32] {
+		storage_prefix(self.pallet_name, self.storage_name)
+	}
+}
+
+impl Debug for StoragePrefix {
+	fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+		write!(
+			f,
+			"{}::{}",
+			sp_std::str::from_utf8(self.pallet_name).unwrap_or(""),
+			sp_std::str::from_utf8(self.storage_name).unwrap_or("")
+		)
+	}
+}
+
+/// A guard that asserts that a specific storage prefix has been mutated or not.
+#[derive(Default, Debug)]
+pub struct StateDiffGuard {
+	// Storage prefixes that are expected to change.
+	whitelisted_prefixes: BTreeSet<StoragePrefix>,
```

Comment:
```suggestion
	must_change_prefixes: BTreeSet<StoragePrefix>,
```
The name here is not really expressing that they must *change*. It is also possible to make this more abstract by having a `Map<Prefix, Change>` where `Change` can be `{ Can | Must } x { Change | NotChange }`, but i think its fine for the beginning.